eks/cleanup.go

```diff
@@ -0,0 +1,328 @@
+	logger.Infof("Deleting security group %s", securityGroupID)
+	delSGInput := &ec2.DeleteSecurityGroupInput{
+		GroupId: aws.String(securityGroupID),
+	}
+	_, err = ec2Svc.DeleteSecurityGroup(delSGInput)
+	if err != nil {
+		if awsErr, isAwsErr := err.(awserr.Error); isAwsErr && awsErr.Code() == "InvalidGroup.NotFound" {
+			logger.Infof("Security group %s already deleted.", securityGroupID)
+			return nil
+		}
+		return errors.WithStackTrace(err)
+	}
+	logger.Infof("Successfully deleted security group with name = %s", securityGroupID)
+
+	// Now delete ALB Ingress Controller's security group, if it exists
+	sgResult, err := lookupSecurityGroup(ec2Svc, vpcID, clusterID)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	for _, result := range sgResult.SecurityGroups {
+		groupID := aws.StringValue(result.GroupId)
+		groupName := aws.StringValue(result.GroupName)
+
+		err = deleteDependencies(ec2Svc, groupID)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+
+		input := &ec2.DeleteSecurityGroupInput{
+			GroupId:   aws.String(groupID),
+			GroupName: aws.String(groupName),
+		}
+		_, err := ec2Svc.DeleteSecurityGroup(input)
+		if err != nil {
+			if awsErr, isAwsErr := err.(awserr.Error); isAwsErr && awsErr.Code() == "InvalidGroup.NotFound" {
+				logger.Infof("Security group %s already deleted.", groupID)
+				return nil
+			}
+			return errors.WithStackTrace(err)
+		}
+
+		logger.Infof("Successfully deleted security group with name=%s, id=%s", groupID, groupName)
+	}
+
+	return nil
+}
+
+// Used to detach and delete elastic network interfaces used by the security group
+// so that the security group can be deleted.
+func deleteDependencies(ec2Svc *ec2.EC2, securityGroupID string) error {
+	logger := logging.GetProjectLogger()
+
+	// Find network interfaces for AWS-managed security group
+	describeNetworkInterfacesInput := &ec2.DescribeNetworkInterfacesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("group-id"),
+				Values: []*string{aws.String(securityGroupID)},
+			},
+		},
+	}
+	networkInterfacesResult, err := ec2Svc.DescribeNetworkInterfaces(describeNetworkInterfacesInput)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	for _, ni := range networkInterfacesResult.NetworkInterfaces {
+		logger.Infof("Found network interface %s", aws.StringValue(ni.NetworkInterfaceId))
+	}
+
+	// Detach network interfaces
+	for _, ni := range networkInterfacesResult.NetworkInterfaces {
+		// First check the network interface has an attachment. It might have gotten detached before we can even process it.
+		// If it doesn't have an attachment, continue to the next network interface.
+		if ni.Attachment == nil || aws.StringValue(ni.Attachment.Status) == "detached" {
+			logger.Infof("Network interface %s is detached.", aws.StringValue(ni.NetworkInterfaceId))
+			continue
+		}
+
+		detachInput := &ec2.DetachNetworkInterfaceInput{
+			AttachmentId: ni.Attachment.AttachmentId,
```

Comment:
✋ I have a question! To make this safe, and not run into a nil pointer deference panic, does this work?

```go
AttachmentId: aws.String(aws.StringValue(ni.Attachment.AttachmentId)),
```

Reply:
I don't think that helps because the issue is that `ni.Attachment` is `nil`, so what you need to do is guard against that.